File: windows/Pcci/Services/EngineClient.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Threading.Tasks;
using Pcci.Models;

namespace Pcci.Services;

/// <summary>
/// Talks to the pcci engine, which ships beside the app as engine\pcci.exe.
///
/// The contract is the engine's: with --json, stdout carries exactly one JSON object
/// and stderr carries JSON Lines. Nothing here parses human-readable text, and every
/// call is awaited off the UI thread.
/// </summary>
public sealed class EngineClient
{
    private readonly string _executable;

    public List<EngineLogLine> Log { get; } = new();

    public EngineClient(string? executable = null)
    {
        _executable = executable ?? FindEngine()
            ?? throw new EngineException(EngineError.Local(
                "The conversion engine is missing from this copy of PCCI.",
                "no pcci.exe found beside the application or on PATH"));
    }

    /// <summary>engine\pcci.exe beside the app, a development build, or PATH.</summary>
    public static string? FindEngine()
    {
        var baseDirectory = AppContext.BaseDirectory;
        var candidates = new[]
        {
            Path.Combine(baseDirectory, "engine", "pcci.exe"),
            Path.Combine(baseDirectory, "engine", "pcci"),
            Path.Combine(baseDirectory, "pcci.exe"),
            Path.GetFullPath(Path.Combine(baseDirectory, "..", "..", "..", "..", "..", "build", "engine", "pcci", "pcci.exe"))
        };
        foreach (var candidate in candidates)
        {
            if (File.Exists(candidate)) return candidate;
        }

        var path = Environment.GetEnvironmentVariable("PATH") ?? "";
        foreach (var directory in path.Split(Path.PathSeparator))
        {
            if (string.IsNullOrWhiteSpace(directory)) continue;
            var candidate = Path.Combine(directory, "pcci.exe");
            if (File.Exists(candidate)) return candidate;
        }
        return null;
    }

    public Task<JsonNode?> DoctorAsync() => RunJsonAsync(new[] { "doctor", "--json" });

    public async Task<SlidePlan> PlanAsync(string sourcePath, EngineConfig config)
    {
        var arguments = new List<string> { "plan", sourcePath, "--json" };
        arguments.AddRange(ConfigArguments(config));
        return DecodePlan(await RunAsync(arguments));
    }

    /// <summary>
    /// Re-plan from a song the user corrected. The chunking rules stay in the engine so
    /// the preview and the exported file can never disagree.
    /// </summary>
    public async Task<SlidePlan> PlanAsync(Song song, EngineConfig config)
    {
        var temporary = WriteTemporary(JsonSerializer.SerializeToUtf8Bytes(song), "song");
        try
        {
            var arguments = new List<string> { "plan", "--song", temporary, "--json" };
            arguments.AddRange(ConfigArguments(config));
            return DecodePlan(await RunAsync(arguments));
        }
        finally
        {
            TryDelete(temporary);
        }
    }

    // Songs from the web. Each of these ends at an ordinary file on disk, which the
    // existing Plan/Build calls then work on exactly as they would on a dropped file.

    /// <summary>
    /// Search every source. The artist is a filter and part of the question both: a
    /// bare title is a thousand songs on any source, a title with an artist is one.
    /// </summary>
    public async Task<SearchOutcome> SearchAsync(
        string query, int limit = 20, string artist = "", string album = "", string year = "")
    {
        var arguments = new List<string> { "search", query, "--limit", limit.ToString(), "--json" };
        if (!string.IsNullOrWhiteSpace(artist)) { arguments.Add("--artist"); arguments.Add(artist); }
        if (!string.IsNullOrWhiteSpace(album)) { arguments.Add("--album"); arguments.Add(album); }
        if (!string.IsNullOrWhiteSpace(year)) { arguments.Add("--year"); arguments.Add(year); }
        var output = await RunAsync(arguments);
        return JsonSerializer.Deserialize<SearchOutcome>(output)
            ?? throw new EngineException(EngineError.Local("The search returned nothing readable."));
    }

    /// <summary>Download the chart at a link. The engine picks where to keep it.</summary>
    public async Task<ImportedChart> FetchAsync(string url)
    {
        var output = await RunAsync(new[] { "fetch", url, "--json" });
        return JsonSerializer.Deserialize<ImportedChart>(output)
            ?? throw new EngineException(EngineError.Local("The download returned nothing readable."));
    }

    /// <summary>
    /// Hand the engine text copied from somewhere else.
    ///
    /// It goes over stdin rather than a temporary file so a chart copied out of an
    /// email never lands on disk as a file nobody asked for; the engine writes it out
    /// itself, once, under a name taken from the song.
    /// </summary>
    public async Task<ImportedChart> PasteAsync(string text)
    {
        var output = await RunAsync(new[] { "paste", "--json" }, input: text);
        return JsonSerializer.Deserialize<ImportedChart>(output)
            ?? throw new EngineException(EngineError.Local("The pasted chart could not be read."));
    }

    public async Task<ConversionResult> BuildAsync(SlidePlan plan, string destination, bool writeChordPro)
    {
        var payload = Encoding.UTF8.GetBytes(plan.Raw?.ToJsonString() ?? "{}");
        var temporary = WriteTemporary(payload, "plan");
        try
        {
            var arguments = new List<string>
            {
                "build", "--plan", temporary, "-o", destination, "--json"
            };
            if (writeChordPro) arguments.Add("--chordpro");
            var output = await RunAsync(arguments);
            return JsonSerializer.Deserialize<ConversionResult>(output)
                ?? throw new EngineException(EngineError.Local("The engine returned nothing."));
        }
        finally
        {
            TryDelete(temporary);
        }
    }

    private static IEnumerable<string> ConfigArguments(EngineConfig config) => new[]
    {
        "-n", config.LinesPerSlide.ToString(),
        "--target", config.ExportTarget,
        config.BalanceLastSlide ? "--balance" : "--no-balance",
        "--chords", config.ChordDelivery,
        "--chord-placement", config.ChordPlacement,
        config.ChordsOnSlide ? "--chords-on-slide" : "--no-chords-on-slide"
    };

    public static SlidePlan DecodePlan(string json)
    {
        var root = JsonNode.Parse(json)
            ?? throw new EngineException(EngineError.Local("The engine returned an empty plan."));
        return new SlidePlan
        {
            Song = root["song"].Deserialize<Song>() ?? new Song(),
            Slides = root["slides"].Deserialize<List<PlannedSlide>>() ?? new List<PlannedSlide>(),
            Config = root["config"].Deserialize<EngineConfig>() ?? new EngineConfig(),
            Warnings = root["warnings"].Deserialize<List<string>>() ?? new List<string>(),
            Raw = root
        };
    }

    private async Task<JsonNode?> RunJsonAsync(IEnumerable<string> arguments) =>
        JsonNode.Parse(await RunAsync(arguments));

    private async Task<string> RunAsync(IEnumerable<string> arguments, string? input = null)
    {
        Log.Clear();
        var info = new ProcessStartInfo
        {
            FileName = _executable,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            RedirectStandardInput = input is not null,
            UseShellExecute = false,
            CreateNoWindow = true,
            StandardOutputEncoding = Encoding.UTF8,
            StandardErrorEncoding = Encoding.UTF8,
            // Explicitly UTF-8 with no byte-order mark. The default here is the
            // console's code page, which on a Windows machine is whatever ANSI
            // codepage that machine happens to use - and a chart pasted from a web
            // page is full of characters that do not survive one.
            StandardInputEncoding = input is null ? null : new UTF8Encoding(false)
        };
        foreach (var argument in arguments)
        {
            info.ArgumentList.Add(argument);
        }

        using var process = new Process { StartInfo = info };
        try
        {
            process.Start();
        }
        catch (Exception exception)
        {
            throw new EngineException(EngineError.Local(
                "PCCI could not start its conversion engine.",
                $"{_executable}: {exception.Message}"));
        }

        var stdoutTask = process.StandardOutput.ReadToEndAsync();
        var stderrTask = process.StandardError.ReadToEndAsync();
        if (input is not null)
        {
            // Written after the readers are started and closed straight after: the
            // engine reads stdin to the end before it answers, so writing it with
            // nobody draining stdout would deadlock on a long chart.
            await process.StandardInput.WriteAsync(input);
            process.StandardInput.Close();
        }
        await process.WaitForExitAsync();
        var stdout = await stdoutTask;
        var stderr = await stderrTask;

        Log.AddRange(ParseLog(stderr));

        if (process.ExitCode != 0)
        {
            throw new EngineException(DecodeError(stdout, process.ExitCode));
        }
        return stdout;
    }

    public static List<EngineLogLine> ParseLog(string stderr)
    {
        var lines = new List<EngineLogLine>();
        foreach (var line in stderr.Split('\n'))
        {
            var trimmed = line.Trim();
            if (trimmed.Length == 0) continue;
            try
            {
                var parsed = JsonSerializer.Deserialize<EngineLogLine>(trimmed);
                if (parsed is not null) lines.Add(parsed);
            }
            catch (JsonException)
            {
                // Not JSON: the contract says logs are JSON Lines, so anything else is
                // noise from a crashing interpreter. Keep it — it is exactly what a bug
                // report needs.
                lines.Add(new EngineLogLine { Level = "raw", Message = trimmed });
            }
        }
        return lines;
    }

    public static EngineError DecodeError(string stdout, int exitCode)
    {
        try
        {
            var envelope = JsonSerializer.Deserialize<EngineErrorEnvelope>(stdout);
            if (envelope?.Error is { } error) return error;
        }
        catch (JsonException)
        {
            // fall through to the generic message
        }
        return EngineError.Local(
            "The conversion did not finish.",
            $"engine exited with status {exitCode}");
    }

    private static string WriteTemporary(byte[] payload, string prefix)
    {
        var path = Path.Combine(Path.GetTempPath(), $"pcci-{prefix}-{Guid.NewGuid():N}.json");
        File.WriteAllBytes(path, payload);
        return path;
    }

    private static void TryDelete(string path)
    {
        try
        {
            File.Delete(path);
        }
        catch (IOException)
        {
            // A leftover temp file is not worth failing a conversion over.
        }
    }
}

File: macos/Sources/PCCI/Engine/EngineClient.swift
import Foundation

/// Talks to the `pcci` engine, which ships inside the app bundle as a sidecar binary.
///
/// The contract is the one the engine documents: with `--json`, stdout carries exactly
/// one JSON object and stderr carries JSON Lines. Nothing here parses human-readable
/// text, and nothing here runs on the main actor — every call is `async` and the
/// process is waited on off the main thread.
actor EngineClient {

    /// Log lines the engine wrote to stderr during the most recent call.
    private(set) var log: [EngineLogLine] = []

    private let executableURL: URL

    init(executableURL: URL? = nil) throws {
        if let executableURL {
            self.executableURL = executableURL
            return
        }
        guard let bundled = EngineClient.bundledEngineURL() else {
            throw EngineError.local(
                "The conversion engine is missing from this copy of PCCI.",
                detail: "no engine found in Contents/Resources/engine or on PATH"
            )
        }
        self.executableURL = bundled
    }

    /// `Contents/Resources/engine/pcci`, or a development build, or `pcci` on PATH.
    static func bundledEngineURL() -> URL? {
        let candidates: [URL?] = [
            Bundle.main.resourceURL?.appendingPathComponent("engine/pcci"),
            Bundle.main.resourceURL?.appendingPathComponent("engine/pcci.exe"),
            URL(fileURLWithPath: "../core/.venv/bin/pcci"),
            URL(fileURLWithPath: "/usr/local/bin/pcci"),
            URL(fileURLWithPath: "/opt/homebrew/bin/pcci")
        ]
        for case let candidate? in candidates
        where FileManager.default.isExecutableFile(atPath: candidate.path) {
            return candidate
        }
        return nil
    }

    // MARK: Commands

    func doctorJSON() async throws -> Data {
        try await runRaw(["doctor", "--json"])
    }

    /// Every call returns raw JSON. Decoding happens on the caller's side, which keeps
    /// values that are not `Sendable` — a decoded plan holds the engine's own JSON
    /// dictionary — from crossing the actor boundary.
    func analyzeJSON(_ url: URL) async throws -> Data {
        try await runRaw(["analyze", url.path, "--json"])
    }

    func planJSON(_ url: URL, config: EngineConfig) async throws -> Data {
        try await runRaw(["plan", url.path, "--json"] + configArguments(config))
    }

    /// Re-plan from a song the user has corrected in the review screen.
    ///
    /// The chunking rules stay in the engine: the app sends the corrected song back
    /// rather than re-implementing slide planning in Swift, so the two can never drift.
    func planJSON(song: Data, config: EngineConfig) async throws -> Data {
        let temporary = try Self.writeTemporary(song, prefix: "pcci-song")
        defer { try? FileManager.default.removeItem(at: temporary) }
        return try await runRaw(
            ["plan", "--song", temporary.path, "--json"] + configArguments(config)
        )
    }

    // MARK: Songs from the web

    /// Search every online source the engine knows.
    ///
    /// The artist is a filter *and* part of the question: a bare title is a thousand
    /// songs on any source, and a title with an artist is usually one.
    func searchJSON(
        _ query: String,
        limit: Int = 20,
        artist: String = "",
        album: String = "",
        year: String = ""
    ) async throws -> Data {
        var arguments = ["search", query, "--limit", String(limit), "--json"]
        if !artist.isEmpty { arguments += ["--artist", artist] }
        if !album.isEmpty { arguments += ["--album", album] }
        if !year.isEmpty { arguments += ["--year", year] }
        return try await runRaw(arguments)
    }

    /// Download the chart at a link. The engine picks where to keep it and says where.
    func fetchJSON(url: String) async throws -> Data {
        try await runRaw(["fetch", url, "--json"])
    }

    /// Hand the engine text copied from somewhere else.
    ///
    /// It goes over stdin rather than a temporary file so a chart the user copied out
    /// of an email never touches the disk as a file nobody asked for; the engine
    /// writes it out itself, once, under a name taken from the song.
    func pasteJSON(text: String) async throws -> Data {
        try await runRaw(["paste", "--json"], input: Data(text.utf8))
    }

    /// Build from a plan the user has reviewed and possibly edited.
    func buildJSON(plan: Data, to destination: URL, writeChordPro: Bool) async throws -> Data {
        let temporary = try Self.writeTemporary(plan, prefix: "pcci-plan")
        defer { try? FileManager.default.removeItem(at: temporary) }
        var arguments = ["build", "--plan", temporary.path, "-o", destination.path, "--json"]
        if writeChordPro { arguments.append("--chordpro") }
        return try await runRaw(arguments)
    }

    private static func writeTemporary(_ payload: Data, prefix: String) throws -> URL {
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("\(prefix)-\(UUID().uuidString).json")
        try payload.write(to: url)
        return url
    }

    private func configArguments(_ config: EngineConfig) -> [String] {
        [
            "-n", String(config.linesPerSlide),
            config.balanceLastSlide ? "--balance" : "--no-balance",
            "--chords", config.chordDelivery.rawValue,
            "--chord-placement", config.chordPlacement.rawValue
        ]
    }

    // MARK: Plan decoding

    /// Decode a plan, keeping the original JSON so an edited plan round-trips whole.
    static func decodePlan(_ data: Data) throws -> SlidePlan {
        guard let raw = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw EngineError.local(
                "The engine returned something PCCI could not read.",
                detail: "plan payload was not a JSON object"
            )
        }
        let decoder = JSONDecoder()
        let songData = try JSONSerialization.data(withJSONObject: raw["song"] ?? [:])
        let slidesData = try JSONSerialization.data(withJSONObject: raw["slides"] ?? [])
        let configData = try JSONSerialization.data(withJSONObject: raw["config"] ?? [:])
        return SlidePlan(
            song: try decoder.decode(Song.self, from: songData),
            slides: try decoder.decode([PlannedSlide].self, from: slidesData),
            config: try decoder.decode(EngineConfig.self, from: configData),
            warnings: raw["warnings"] as? [String] ?? [],
            raw: raw
        )
    }

    // MARK: Decoding

    /// Decode any of the engine's payloads, saying which one failed rather than
    /// letting a `DecodingError` reach the user as a wall of Swift.
    static func decode<T: Decodable>(_ type: T.Type, from data: Data, what: String) throws -> T {
        do {
            return try JSONDecoder().decode(type, from: data)
        } catch {
            throw EngineError.local(
                "PCCI could not read what the engine said about \(what).",
                detail: "\(error)"
            )
        }
    }

    // MARK: Process plumbing

    /// Run the engine and return stdout, turning a non-zero exit into a typed error.
    private func runRaw(_ arguments: [String], input: Data? = nil) async throws -> Data {
        log.removeAll()
        let process = Process()
        process.executableURL = executableURL
        process.arguments = arguments

        let outPipe = Pipe()
        let errPipe = Pipe()
        process.standardOutput = outPipe
        process.standardError = errPipe
        let inPipe = input.map { _ in Pipe() }
        if let inPipe { process.standardInput = inPipe }

        do {
            try process.run()
        } catch {
            throw EngineError.local(
                "PCCI could not start its conversion engine.",
                detail: "\(executableURL.path): \(error.localizedDescription)"
            )
        }

        // Written off the calling thread and closed straight after: the engine reads
        // stdin to the end before it answers, so writing it inline would deadlock as
        // soon as the text outgrew a pipe buffer.
        if let inPipe, let input {
            DispatchQueue.global(qos: .userInitiated).async {
                inPipe.fileHandleForWriting.write(input)
                try? inPipe.fileHandleForWriting.close()
            }
        }

        let outData = try await Self.readToEnd(outPipe)
        let errData = try await Self.readToEnd(errPipe)
        await Self.wait(for: process)

        log = Self.parseLog(errData)

        guard process.terminationStatus == 0 else {
            throw Self.decodeError(from: outData, exitCode: process.terminationStatus, log: log)
        }
        return outData
    }

    private static func readToEnd(_ pipe: Pipe) async throws -> Data {
        try await withCheckedThrowingContinuation { continuation in
            DispatchQueue.global(qos: .userInitiated).async {
                let data = pipe.fileHandleForReading.readDataToEndOfFile()
                continuation.resume(returning: data)
            }
        }
    }

    private static func wait(for process: Process) async {
        await withCheckedContinuation { continuation in
            DispatchQueue.global(qos: .userInitiated).async {
                process.waitUntilExit()
                continuation.resume()
            }
        }
    }

    static func parseLog(_ data: Data) -> [EngineLogLine] {
        String(decoding: data, as: UTF8.self)
            .split(separator: "\n")
            .compactMap { line in
                guard
                    let payload = try? JSONSerialization.jsonObject(
                        with: Data(line.utf8)
                    ) as? [String: Any]
                else {
                    return nil
                }
                return EngineLogLine(
                    level: payload["level"] as? String ?? "info",
                    message: payload["msg"] as? String ?? "",
                    timestamp: payload["ts"] as? String ?? ""
                )
            }
    }

    static func decodeError(from data: Data, exitCode: Int32, log: [EngineLogLine]) -> EngineError {
        if
            let payload = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
            let error = payload["error"],
            let errorData = try? JSONSerialization.data(withJSONObject: error),
            let decoded = try? JSONDecoder().decode(EngineError.self, from: errorData)
        {
            return decoded
        }
        let tail = log.suffix(3).map(\.message).joined(separator: "; ")
        return EngineError.local(
            "The conversion did not finish.",
            detail: "engine exited with status \(exitCode). \(tail)"
        )
    }
}
